libgalois/include/katana/PropertyGraphRetractor.h

```diff
@@ -69,10 +69,30 @@ class KATANA_EXPORT PropertyGraphRetractor {
     return pg_->rdg_.edge_properties();
   }
 
+  /// Return true if type information has been loaded separate from properties.
+  /// Return false otherwise.
+  bool IsEntityTypeIDsOutsideProperties() {
+    return pg_->rdg_.IsEntityTypeIDsOutsideProperties();
+  }
+
+  /// Modifiable access to the underlying type array.
+  /// This is exposed because type id mappings change sometimes.
+  PropertyGraph::EntityTypeIDArray& NodeTypeArray() {
+    return pg_->node_entity_type_ids_;
+  }
+
+  /// Modifiable access to the underlying type array.
+  /// This is exposed because type id mappings change sometimes.
+  PropertyGraph::EntityTypeIDArray& EdgeTypeArray() {
+    return pg_->edge_entity_type_ids_;
+  }
+
+  /// This is exposed because type id mappings change sometimes.
   void ReplaceNodeTypeManager(EntityTypeManager&& manager) {
     pg_->node_entity_type_manager_ = std::move(manager);
```

Comment:
I wonder if we should have a valid bit in the EntityTypeManager, so if you use one after it being moved, it barfs all over you.

Reply:
We already have an `Init()` so there is a logical place to set the bit. I would be on board.

Reply:
I'm on it.

Reply:
I personally am not a fan of valid bit but I realize they are useful sometimes. valid bits are good until they aren't. IMO, it's better to deduce validity through the state of the object, e.g. empty maps mean EntityTypeManager is invalid because a valid one would have at least `kUnknownType` mapped.

Reply:
Well, I implemented it in my branch, so let's discuss before it becomes a PR.  What is ugly about the valid bit is that it needs to be checked everywhere.  But it provides fast fail for when an object has been moved.  This happens in graph update, where I move entity type managers and I've certainly coded the bug where I use it after moving it.  But maybe this problem isn't important enough to write and maintain a valid bit.

Amber, no "empty behavior" is sufficient.  I want to fail an assert if I try to use an invalid manager.

Reply:
I'd like to see your code that has these moves and related bugs (or possibility of one). I also still don't understand why we can't implement `isValid()` by checking for empty maps? A "default/no-arg" constructed manager has non-empty maps, and a moved one should have empty maps.

Reply:
Moved discussion to slack.
https://thek-teamworkspace.slack.com/archives/C028THYNA07/p1631379992027200